### ISSUE TITLE
no-unsafe-finally triggers even without control structures in the try { } block

### ISSUE THREAD

**Issue by dandv** [eslint/eslint#11297]
https://eslint.org/docs/rules/no-unsafe-finally states,

> JavaScript suspends the control flow statements of `try` and catch blocks until the execution of `finally`[sic] block finishes. So, when `return`, throw, break, or continue is used in finally, **control flow statements inside try** and catch are overwritten, which is considered as unexpected behavior.

The problem is that this rule triggers even if there are no control flow statements in the `try`. A pattern would be assigning a default value to a variable, then trying to obtain another value from a call that might throw:

```js
(() => {
    let foo = 'default';
    try {
        // No control structures here to be suspended
        foo = 'bar';  // in practice, some call that can throw
    } finally {
        // No need to trigger no-unsafe-finally
        return {
            foo
        }
    }
})();

```
Here's the [eslint.org demo](https://eslint.org/demo/#eyJ0ZXh0IjoiKCgpID0+IHtcbiAgICBsZXQgZm9vID0gJ2RlZmF1bHQnO1xuICAgIHRyeSB7XG4gICAgICAgIC8vIE5vIGNvbnRyb2wgc3RydWN0dXJlcyBoZXJlIHRvIGJlIHN1c3BlbmRlZFxuICAgICAgICBmb28gPSAnYmFyJzsgIC8vIGluIHByYWN0aWNlLCBzb21lIG9wZXJhdGlvbiB0aGF0IGNhbiB0aHJvd1xuICAgIH0gZmluYWxseSB7XG4gICAgICAgIC8vIE5vIG5lZWQgdG8gdHJpZ2dlciBuby11bnNhZmUtZmluYWxseVxuICAgICAgICByZXR1cm4ge1xuICAgICAgICAgICAgZm9vXG4gICAgICAgIH1cbiAgICB9XG59KSgpO1xuIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo5LCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7ImNvbnN0cnVjdG9yLXN1cGVyIjoyLCJmb3ItZGlyZWN0aW9uIjoyLCJnZXR0ZXItcmV0dXJuIjoyLCJuby1jYXNlLWRlY2xhcmF0aW9ucyI6Miwibm8tY2xhc3MtYXNzaWduIjoyLCJuby1jb21wYXJlLW5lZy16ZXJvIjoyLCJuby1jb25kLWFzc2lnbiI6Miwibm8tY29uc29sZSI6Miwibm8tY29uc3QtYXNzaWduIjoyLCJuby1jb25zdGFudC1jb25kaXRpb24iOjIsIm5vLWNvbnRyb2wtcmVnZXgiOjIsIm5vLWRlYnVnZ2VyIjoyLCJuby1kZWxldGUtdmFyIjoyLCJuby1kdXBlLWFyZ3MiOjIsIm5vLWR1cGUtY2xhc3MtbWVtYmVycyI6Miwibm8tZHVwZS1rZXlzIjoyLCJuby1kdXBsaWNhdGUtY2FzZSI6Miwibm8tZW1wdHktY2hhcmFjdGVyLWNsYXNzIjoyLCJuby1lbXB0eS1wYXR0ZXJuIjoyLCJuby1lbXB0eSI6Miwibm8tZXgtYXNzaWduIjoyLCJuby1leHRyYS1ib29sZWFuLWNhc3QiOjIsIm5vLWV4dHJhLXNlbWkiOjIsIm5vLWZhbGx0aHJvdWdoIjoyLCJuby1mdW5jLWFzc2lnbiI6Miwibm8tZ2xvYmFsLWFzc2lnbiI6Miwibm8taW5uZXItZGVjbGFyYXRpb25zIjoyLCJuby1pbnZhbGlkLXJlZ2V4cCI6Miwibm8taXJyZWd1bGFyLXdoaXRlc3BhY2UiOjIsIm5vLW1peGVkLXNwYWNlcy1hbmQtdGFicyI6Miwibm8tbmV3LXN5bWJvbCI6Miwibm8tb2JqLWNhbGxzIjoyLCJuby1vY3RhbCI6Miwibm8tcmVkZWNsYXJlIjoyLCJuby1yZWdleC1zcGFjZXMiOjIsIm5vLXNlbGYtYXNzaWduIjoyLCJuby1zcGFyc2UtYXJyYXlzIjoyLCJuby10aGlzLWJlZm9yZS1zdXBlciI6Miwibm8tdW5kZWYiOjIsIm5vLXVuZXhwZWN0ZWQtbXVsdGlsaW5lIjoyLCJuby11bnJlYWNoYWJsZSI6Miwibm8tdW5zYWZlLWZpbmFsbHkiOjIsIm5vLXVuc2FmZS1uZWdhdGlvbiI6Miwibm8tdW51c2VkLWxhYmVscyI6Miwibm8tdW51c2VkLXZhcnMiOjIsIm5vLXVzZWxlc3MtZXNjYXBlIjoyLCJyZXF1aXJlLXlpZWxkIjoyLCJ1c2UtaXNuYW4iOjIsInZhbGlkLXR5cGVvZiI6MiwiYXJyYXktYnJhY2tldC1zcGFjaW5nIjoyLCJkZWZhdWx0LWNhc2UiOjIsImluZGVudCI6Mn0sImVudiI6eyJub2RlIjp0cnVlLCJwaGFudG9tanMiOmZhbHNlfX19).

**Comment by not-an-aardvark** [comment]
Hi, thanks for the report. Perhaps the rule's description could be worded better -- I think "control flow statements inside try" is intended to mean "return, break, or continue in `try`, or something in `try` that throws an error".

For example, if your code looks like this:

```js
(() => {
    let foo = 'default';
    try {
        foo = maybeThrows();
    } finally {
        return {
            foo
        }
    }
})();
```

Then if `maybeThrows()` throws an error, the error will be swallowed and `{ foo: 'default' }` will be returned. This is potentially confusing because `finally` is typically used to perform some kind of cleanup rather than to swallow errors (if you want to swallow errors, `catch` would be a better option).

In other words, I think the rule is working as intended by reporting a frequently-buggy pattern, although in this case the pattern might not actually indicate the presence of a bug.

**Comment by dandv** [comment]
I agree that an empty `catch` would be more idiomatic. I was confused because `foo = 'bar'` cannot throw.

**Comment by not-an-aardvark** [comment]
To clarify, is there a use case for using `try` around things that can't throw? I think the rule doesn't check whether the contents of the `try` block can throw, because if they can't throw you would be better off just omitting the `try` block.

**Comment by dandv** [comment]
No use case really. I agree that this isn't an wait error really, and the
docs could be improved as described in the second comment.
